---
# Teoría de Probabilidades


### Espacio Muestral

El espacio muestral, denotado típicamente como $S$, es el conjunto que contiene todos los posibles resultados individuales de un experimento aleatorio. Es decir, es el conjunto de todos los resultados que podrían ocurrir al realizar un experimento. Por ejemplo, si lanzas un dado, el espacio muestral sería el conjunto $S = \left \{ 1,2,3,4,5,6\right \}$, que incluye los seis posibles resultados del lanzamiento del dado.


Suponga que se cuenta con una moneda justa, con opciones cara y sello. Se realiza el experimento de lanzamiento de la moneda dos veces. ¿Cuál es el espacio muestral?

In [ ]:
posibilidades = {'cara', 'sello'}
espacio_muestral = set()

for tiro1 in posibilidades:
    for tiro2 in posibilidades:
        resultado = (tiro1, tiro2)
        espacio_muestral.add(resultado)

print(espacio_muestral)

{('sello', 'sello'), ('cara', 'cara'), ('sello', 'cara'), ('cara', 'sello')}


¿Cuál sería el espacio muestral de un experimento que consiste en lanzar 2 veces un dado justo de 6 caras?

In [ ]:
posibilidades = {1,2,3,4,5,6}
espacio_muestral = set()

for tiro1 in posibilidades:
    for tiro2 in posibilidades:
        resultado = (tiro1, tiro2)
        espacio_muestral.add(resultado)

print(espacio_muestral)
print(len(espacio_muestral))


{(3, 4), (4, 3), (3, 1), (5, 4), (4, 6), (5, 1), (2, 2), (1, 6), (2, 5), (1, 3), (6, 2), (6, 5), (4, 2), (4, 5), (3, 3), (5, 6), (3, 6), (5, 3), (2, 4), (1, 2), (2, 1), (1, 5), (6, 1), (6, 4), (3, 2), (4, 1), (3, 5), (5, 2), (4, 4), (5, 5), (1, 1), (1, 4), (2, 3), (2, 6), (6, 6), (6, 3)}
36


¿Qué pasaría con el código si el experimento consiste en lanzar una mayor cantidad de veces el dado? Deberíamos anidar más ciclos, pero eso es ineficiente. Acá hay una forma más eficiente.

In [ ]:
from itertools import product

In [ ]:
posibilidades = {'cara', 'sello'}
espacio_muestral = set(product(posibilidades, repeat=2))

print(espacio_muestral)

{('sello', 'sello'), ('cara', 'cara'), ('sello', 'cara'), ('cara', 'sello')}


¿y si hacemos 3 lanzamientos? ¿cuál sería el espacio muestral?

In [ ]:
posibilidades = {'cara', 'sello'}
espacio_muestral = set(product(posibilidades, repeat=3))

print(espacio_muestral)

{('sello', 'cara', 'cara'), ('cara', 'sello', 'cara'), ('sello', 'sello', 'cara'), ('cara', 'cara', 'sello'), ('sello', 'cara', 'sello'), ('cara', 'cara', 'cara'), ('cara', 'sello', 'sello'), ('sello', 'sello', 'sello')}


### Eventos

Un evento es cualquier subconjunto del espacio muestral, es decir, es cualquier colección de resultados posibles de un experimento. Los eventos pueden ser simples (constituidos por un solo resultado) o compuestos (constituidos por más de un resultado). Por ejemplo, si lanzas un dado, algunos eventos posibles podrían ser:

- Obtener un número par: $E_{1} = \left \{ 2,4,6 \right \}$
- Obtener un número impar: $E_{2} = \left \{ 1,3,5 \right \}$
- Obtener un número mayor que 4: $E_{3} = \left \{ 5,6 \right \}$


Los elementos que integran a un evento son aquellos elementos del espacio muestral que satisfacen una condición C determinada, que en notación formal se escribe:

$E = \left \{ \forall x \in \Omega \mid x \quad satisface \quad C \right \}$


In [ ]:
# En un lanzamiento de dados, enumere el evento E = {valor menor que 4}

espacio_muestral = set(range(1,7))
E = set()
for posibilidad in espacio_muestral:
    if posibilidad < 4:
        E.add(posibilidad)

print(E)

{1, 2, 3}


In [ ]:
# este mismo código se puede optimizar utilizando list comprehension
espacio_muestral = set(range(1,7))
E = set( [x for x in espacio_muestral if x < 4] )

print('Espacio Muestral', espacio_muestral)
print('Evento', E)

Espacio Muestral {1, 2, 3, 4, 5, 6}
Evento {1, 2, 3}


### Enfoque clásico de cálculo de Probabilidad

En palabras simples, para calcular la probabilidad de un evento, se divide el número de resultados favorables (elementos del evento) por el número total de resultados posibles en el espacio muestral. Es decir:

$P(E) = \frac{\#(E)}{\#(\Omega )}$

En donde,
- $P(E)$ es la probabilidad del evento
- $\#(E)$ es el número de elementos en el evento
- $\#(\Omega)$ es el número de elementos en el espacio muestral $S$.



In [ ]:
# Probabilidad de lanzar un dado y obtener un numero menor de 4
espacio_muestral = set(range(1,7))
E = set( [x for x in espacio_muestral if x < 4] )

P_evento = len(E) / len(espacio_muestral)

print("La probabilidad es", P_evento)

La probabilidad es 0.5


In [ ]:
# probabilidad de sumar 7 en un lanzamiento de dos dados
posibilidades = {1,2,3,4,5,6}
espacio_muestral = set(product(posibilidades, repeat=2))

print("Espacio Muestral", espacio_muestral)

E = [x for x in espacio_muestral if sum(x) == 7]

print('Evento', E)

P_evento = len(E) / len(espacio_muestral)

print("La probabilidad es", P_evento)


Espacio Muestral {(3, 4), (4, 3), (3, 1), (5, 4), (4, 6), (5, 1), (2, 2), (1, 6), (2, 5), (1, 3), (6, 2), (6, 5), (4, 2), (4, 5), (3, 3), (5, 6), (3, 6), (5, 3), (2, 4), (1, 2), (2, 1), (1, 5), (6, 1), (6, 4), (3, 2), (4, 1), (3, 5), (5, 2), (4, 4), (5, 5), (1, 1), (1, 4), (2, 3), (2, 6), (6, 6), (6, 3)}
Evento [(3, 4), (4, 3), (1, 6), (2, 5), (6, 1), (5, 2)]
La probabilidad es 0.16666666666666666


### Enfoque Frecuentista para calcular la probabilidad

La probabilidad frecuencial, frecuentista o empírica es la probabilidad de que un evento ocurra en un número grande de ensayos. Se calcula como el cociente entre el número de ocurrencias de un evento, y el número total de ensayos en un experimento real.

In [ ]:
import random

In [ ]:
# cálculo de probabilidad frecuentista en el lanzamiento de una moneda

d = ['cara', 'sello']
N_evento = 0
n = 10000

for i in range(n):
    if random.choice(d) == 'cara':
        N_evento += 1

print('Probabilidad Cara: ', N_evento/n)

Probabilidad Cara:  0.4937


In [ ]:
# y si fuera un dado, ¿cual seria la probabilidad de que saliera el 6?
d = [1,2,3,4,5,6]
N_evento = 0
n = 10000

for i in range(n):
    if random.choice(d) == 6:
        N_evento += 1

print('Probabilidad Cara: ', N_evento/n)


Probabilidad Cara:  0.1719


---

### Operaciones con Eventos

En teoría de la probabilidad, las operaciones de eventos son acciones que se pueden realizar sobre eventos para crear nuevos eventos. Estas operaciones son similares a las operaciones conjuntistas que se realizan sobre conjuntos en teoría de conjuntos.

- Unión de eventos
- Intersección de eventos
- Complemento de Eventos

Tomemos un dado justo y definamos los siguientes eventos:

- A: obtener un número par
- B: obtener un número mayor que 3



In [ ]:
# calcular la probabilidad de obtener un número par o mayor de 3

espacio_muestral = set(range(1,7))
E = set( [x for x in espacio_muestral if x > 3 or x%2==0] )

P_evento = len(E) / len(espacio_muestral)

print("La probabilidad es", P_evento)


# usando sets
espacio_muestral = {1,2,3,4,5,6}
A = {2,4,6}
B = {4,5,6}

print('Evento A o B', A | B)
print('P_evento', len(A | B)/len(espacio_muestral))


La probabilidad es 0.6666666666666666
Evento A o B {2, 4, 5, 6}
P_evento 0.6666666666666666


In [ ]:
# calcular la probabilidad de obtener un número par mayor de 3

espacio_muestral = set(range(1,7))
E = set( [x for x in espacio_muestral if x > 3 and x%2==0] )

P_evento = len(E) / len(espacio_muestral)

print("La probabilidad es", P_evento)

# usando sets
espacio_muestral = {1,2,3,4,5,6}
A = {2,4,6}
B = {4,5,6}

print('Evento A y B', A & B)
print('P_evento', len(A & B)/len(espacio_muestral))


La probabilidad es 0.3333333333333333
Evento A y B {4, 6}
P_evento 0.3333333333333333


In [ ]:
# calcular la probabilidad de NO obtener un número par mayor de 3

espacio_muestral = set(range(1,7))
E = set( [x for x in espacio_muestral if not (x > 3 and x%2==0)] )

P_evento = len(E) / len(espacio_muestral)

print("La probabilidad es", P_evento)

# usando sets
espacio_muestral = {1,2,3,4,5,6}
A = {2,4,6}
B = {4,5,6}

print('Evento No (A y B)', espacio_muestral - (A & B))
print('P_evento', len(espacio_muestral - (A & B))/len(espacio_muestral))

La probabilidad es 0.6666666666666666
Evento No (A y B) {1, 2, 3, 5}
P_evento 0.6666666666666666


#### Probabilidad Marginal

La probabilidad marginal es la probabilidad de que ocurra un solo evento independientemente de otros eventos. Veamos el siguiente ejemplo:

De un grupo de 7 estudiantes que asistena tutorías: 5 asisten a clases de física, 4 a química y 2 a ambas. Determine las probabilidades marginales de que un estudiante estudie física y de que estudie química.

- A : {estudia física}
- B : {estudia química}

P(A) = 5/7

P(B) = 4/7

#### Probabilidad Condicional

La probabilidad condicional es la probabilidad de que ocurra un evento cuando se sabe que ocurre otro evento específico. La notación de la probabilidad condicional es:

$P(E_{1} \mid E_{2})$

En el ejemplo anterior la probabilidad de seleccionar un estudiante que estudie física sabiendo que estudia química es:

$P(A \mid B) = 2/4$ , esto, debido a que los estudiantes que estudian química son 4 y de entre ellos solo 2 estudian física.

$P(B \mid A) = 2/5$

In [ ]:
# En una heladería hay dos sabores: vainilla y limón. Calcule la probabilidad frecuentista de que si un niño pide limón, tambien pida vainilla.

opciones = ['vainilla','limon']

n = 10000
count_limon = 0
count_vainilla = 0
count_limon_vainilla = 0

for i in range(n):
    sabor1 = random.choice(opciones)
    sabor2 = random.choice(opciones)

    if sabor1 == 'limon' or sabor2 == 'limon':
        count_limon += 1

    if (sabor1 == 'limon' and sabor2 == 'vainilla') or (sabor1 == 'vainilla' and sabor2 == 'limon'):
        count_limon_vainilla += 1

P = count_limon_vainilla / count_limon

print(P)




0.6637804715598774


---
#### Teorema de Bayes

Thomas Bayes propuso un método para calcular la probabilidad condicional de un evento $E$ respecto a otro evento
$F$ si se conoce las probabilidades de los eventos y la probabilidad condicional de $F$ sabiendo que ha ocurrido $E$.

$P(E|F) = P(E) \frac{P(F|E)}{P(F)}$


En términos generales, dado un número de eventos que forman un grupo exhaustivo y considerando el teorema de
la probabilidad total, el teorema de Bayes se escribe de la siguiente forma:


$P(E|F) = P(E) \frac{P(F|E)}{ \sum_{i=1}^{n} P(F|E_{i})P(E_{i}) }$


**Ejemplo**

Una fábrica de papas fritas (chips) compra papas a dos proveedores. Las papas deben tener un tamaño
mínimo caso contrario son consideradas de rechazo. En un muestreo se determina que el porcentaje de papas rechazadas del primer proveedor es igual a 20%, mientras que
del segundo proveedor es el 10%. Si se sabe que **al primer proveedor se le compra cinco veces más que al segundo**,
determine la probabilidad de que al encontrar una papa de rechazo en la línea de producción esta sea del segundo proveedor.

Definamos los eventos:
- $E_{1}$ = {papa comprada al primer proveedor}
- $E_{2}$ = {papa comprada al segundo proveedor}
- $R$ = {la papa es de rechazo}

Entonces,

- $P(E_{1})$ es la probabilidad de comprar al primer proveedor. Como primer proveedor se le compra 5 veces más que al segundo, se puede calcular la probabilidad que una papa sea del primer proveedor de la siguiente manera:

Sea $n_{1}$ la cantidad de papas compradas al primer proveedor, entonces:

$P(E_{1}) = \frac{n_{1}}{n_{1} + \frac{1}{5} n_{1}} = \frac{5}{6} = 0.833$

y por lo tanto,

$P(E_{2}) = 1 - P(E_{1}) = 0.167$

Además, del enunciado, se tiene que:

- $P(R|E_{1}) = 0.2$
- $P(R|E_{2}) = 0.1$

Calculamos la probabilidad de obtener una papa de rechazo:

$P(R) = 0.833 \times 0.2 + 0.167 \times 0.1 = 0.183$

Entonces, lo que queremos es encontrar la probabilidad de que si encontramos una papa de rechazo ésta sea del segundo proveedor. O sea, necesitamos calcular:

$P(E_{2}|R)$

Acá podemos utilizar el teorema de Bayes:

$P(E_{2}|R) = P(E_{2}) \frac{P(R|E_{2})}{P(R)} = 0.167 \times \frac{0.1}{0.183} = 0.091$


En conclusión, la probabilida de encontrar una papa de rechazo y ésta sea del segundo proveedor, es de 9.1%





In [ ]:
# Desafío: programe este problema utilizando el enfoque frecuentista






Por último, debemos señalar que el teorema de Bayes generalmente se interpreta en términos de causa - efecto. Los eventos $E_{i}$ se consideran las
causas y el evento $F$ el efecto.

---